be/src/storage/rowset/column_writer.cpp

```diff
@@ -816,4 +871,136 @@ Status StringColumnWriter::check_string_lengths(const Column& column) {
+            return Status::OK();
+        }
+    }
+    _buf_column->append(column, 0, column.size());
+    if (_buf_column->size() < config::dictionary_speculate_min_chunk_size) {
+        return Status::OK();
+    } else {
+        _is_speculated = true;
+        RETURN_IF_ERROR(speculate_column_and_set_encoding(*_buf_column));
+        Status st = _scalar_column_writer->append(*_buf_column);
+        _buf_column.reset();
+        return st;
+    }
+}
+
+inline Status DictColumnWriter::speculate_column_and_set_encoding(const Column& column) {
+    Status st;
+    EncodingTypePB detect_encoding;
+    LogicalType logicalType = delegate_type(type_info()->type());
+    switch (logicalType) {
+    case TYPE_TINYINT:
+        detect_encoding = speculate_encoding<TYPE_TINYINT>(column);
+        break;
+    case TYPE_SMALLINT:
+        detect_encoding = speculate_encoding<TYPE_SMALLINT>(column);
+        break;
+    case TYPE_INT:
+        detect_encoding = speculate_encoding<TYPE_INT>(column);
+        break;
+    case TYPE_BIGINT:
+        detect_encoding = speculate_encoding<TYPE_BIGINT>(column);
+        break;
+    case TYPE_LARGEINT:
+        detect_encoding = speculate_encoding<TYPE_LARGEINT>(column);
+        break;
+    case TYPE_FLOAT:
+        detect_encoding = speculate_encoding<TYPE_FLOAT>(column);
+        break;
+    case TYPE_DOUBLE:
+        detect_encoding = speculate_encoding<TYPE_DOUBLE>(column);
+        break;
+    case TYPE_DATE:
+        detect_encoding = speculate_encoding<TYPE_DATE>(column);
+        break;
+    case TYPE_DATETIME:
+        detect_encoding = speculate_encoding<TYPE_DATETIME>(column);
+        break;
+    case TYPE_DECIMAL:
+        detect_encoding = speculate_encoding<TYPE_DECIMAL>(column);
+        break;
+    case TYPE_DECIMALV2:
+        detect_encoding = speculate_encoding<TYPE_DECIMALV2>(column);
+        break;
+    default:
+        return Status::InternalError(strings::Substitute("$0 type should not use dictionary encoding", logicalType));
+    }
+    st = _scalar_column_writer->set_encoding(detect_encoding);
+    CHECK(st.ok()) << st;
+    return st;
+}
+
+// The detection logic here uses a set to record the distinct values of a sample column. When the number
+// of distinct values exceeds row_count * ratio, dictionary encoding is no longer used.
+// Here, row_count is the number of elements in the sample column, and ratio is set by the user.
+template <LogicalType Type>
+inline EncodingTypePB DictColumnWriter::speculate_encoding(const Column& column) {
+    using ColumnType = typename RunTimeTypeTraits<Type>::ColumnType;
+    const ColumnType* bin_col;
+    if (column.is_nullable()) {
+        const auto& data_col = down_cast<const NullableColumn&>(column).data_column();
+        bin_col = &down_cast<ColumnType&>(*data_col);
+    } else {
+        bin_col = &down_cast<const ColumnType&>(column);
+    }
+    const size_t dictionary_min_rowcount = 256;
```

Comment:
You can place `dictionary_min_rowcount` in the `olap_define.h` to reuse it with `StringColumnWriter`

Reply:
@chaoyli updated